polly tts

JarbasAl opened the pull request:
add polly tts engine option

penrods commented on mycroft/tts/__init__.py:
```diff
@@ -294,11 +294,15 @@ def __init__(self, tts):
         self.validate_lang()
         self.validate_connection()
 
+    def validate_dependencies(self):
```
Is there a reason you implemented it this way instead of just making Boto3 a dependency in the "requirements.txt"?  I can think of a few reasons, just  curious about the why here.

JarbasAl replied:
being a modular component i think it doesnt make sense to install all dependencies for all possible engines by default, we would end with many unused requirements

the user is already making a conscious choice when changing the engine, i think a warning for extra requirements works alright, maybe i could make it install the requirements automatically instead? 

i think this approach should be extended to TTS and Hotword (snowboy also isn't in  requirements.txt)

penrods replied:
I agree with you and think we have had a habit of semi-exposing too many options and sucking in a bunch of rarely used cruft.  For example, we have "ESpeak" as an option in Home although you can't use it unless you have setup a bunch of stuff manually on your system.

So I think I like the "validate_dependencies()" approach for those who are experimenting.  They will be able to see error messages in the logs this way.

domcross commented:
Is there any reason why this PR was not merged? I have a similar PR ready and just found that AWS Polly support is already there (in theory)...

forslund commented:
I'm not sure maybe it's the codacy quality report or some of the discussion up there wasn't resolved. I'll flag this for re-review :)

andlo commented:
I am waiting on this merge to - as soon it is merged I continue working on danish translations, but need a good danish voice and Amazon has a nice danish male voice I like :) 
I have been using this for a while, but got frustrated when updating mycroft-core.

andlo commented:
Anything I can do to get polly TTS added to mycroft core ? Ineed it badly as it is only good danish male voice I can find. 
If this isnt good enough - maybe the work @domcross  has in hand would be ? Or I can reuse this PR and make a new PR based onnewer mycroft-core ?

Please It would be realy helpfull for me to get polly going.

andlo commented:
OK I took the nice work from this PR and updated it to work with latest mycroft and added test for lang and minor changes to.

I make a PR with that to replace this old one.

pep8speaks commented:
Hello @JarbasAl! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`mycroft/tts/polly_tts.py`](https://github.com/MycroftAI/mycroft-core/blob/61b9c459820c65ef3a2ae775c311c36340db4b3f/mycroft/tts/polly_tts.py):

> [Line 51:42](https://github.com/MycroftAI/mycroft-core/blob/61b9c459820c65ef3a2ae775c311c36340db4b3f/mycroft/tts/polly_tts.py#L51): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\w'

JarbasAl commented:
rebased and updated, test with
```python
    e = PollyTTS()
    e.voice = "Matthew"
    ssml = """<speak>
     This is my original voice, without any modifications. <amazon:effect vocal-tract-length="+15%"> 
     Now, imagine that I am much bigger. </amazon:effect> <amazon:effect vocal-tract-length="-15%"> 
     Or, perhaps you prefer my voice when I'm very small. </amazon:effect> You can also control the 
     timbre of my voice by making minor adjustments. <amazon:effect vocal-tract-length="+10%"> 
     For example, by making me sound just a little bigger. </amazon:effect><amazon:effect 
     vocal-tract-length="-10%"> Or, making me sound only somewhat smaller. </amazon:effect> 
</speak>"""
    e.get_tts(ssml, "polly.mp3")
```

forslund commented:
Works well :+1: Question, would it be clearer if the keys were named `access_key` and `secret_key` in the config so they're more similar?

forslund commented:
Talking with @JarbasAI we figured that the entries should match the aws key names as best as possible. It's Access Key ID and Secret Access Key on aws so the names should probably be `access_key_id` and `secret_access_key`

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-1262))

forslund commented:
Thank you very much for this @JarbasAI, Merging

krisgesling commented:
Thanks Jarbas!

Does the following sound right for the docs? Particularly the AWS account setup, I haven't actually gone through it myself.

## Amazon Polly

[Amazon Polly's](https://aws.amazon.com/polly/) text-to-speech service.

### Account Setup

[Create an AWS account](https://portal.aws.amazon.com/billing/signup?nc2=h_ct&src=default&redirect_url=https%3A%2F%2Faws.amazon.com%2Fregistration-confirmation#/start) and add the Polly service. You will need to take note of your private "Access Key ID" and "Secret Access Key".

### Mycroft Configuration

Check the [list of available voices and languages](https://docs.aws.amazon.com/polly/latest/dg/voicelist.html).

Using the [Configuration Manager](config-manager.md) we can edit the `mycroft.conf` file by running:

```bash
mycroft-config edit user
```

To our existing configuration values we will add the following:

```javascript
"tts": {
  "module": "polly",
  "polly": {
    "voice": "Matthew",
    "region": "us-east-1",
    "access_key_id": "YOUR_ACCESS_KEY_ID",
    "secret_access_key": "YOUR_SECRET_ACCESS_KEY"
  }
}
```

If the `voice` and `region` attributes are ommitted the defaults of `Matthew` and `us-east-1` will be used.

forslund commented:
@krisgesling also to note, the extra python module dependency boto3 needs to be installed

`mycroft-pip install boto3`

JarbasAl commented:
@krisgesling maybe in the docs point to the polly docs section on voices. add a note that language is derived by voice name, it does not respect the global "lang" parameter nor expects it's own like other engines

krisgesling commented:
Did we not merge this? Or is there another PR I'm missing?

devs-mycroft commented:
Um yeah, I thought this was merged... What happened here.

JarbasAl commented:
wut, this was merged, code is here https://github.com/MycroftAI/mycroft-core/blob/dev/mycroft/tts/polly_tts.py

whats up with github?

forslund commented:
Yeah but there was a snafu, where it wasn't for a while

krisgesling commented:
Yeah was my bad, I had it in draft release notes and didn't verify that it actually got merged.

Release notes are now written only from the commit history so until something is merged, it doesn't get in the notes.
